File: extras/rhubarb-for-spine/rhubarb-for-spine/MouthNaming.cs
using System.Collections.Generic;
using System.Linq;

namespace rhubarb_for_spine {
	public class MouthNaming {
		public MouthNaming(string prefix, string suffix, MouthShapeCasing mouthShapeCasing) {
			Prefix = prefix;
			Suffix = suffix;
			MouthShapeCasing = mouthShapeCasing;
		}

		public string Prefix { get; }
		public string Suffix { get; }
		public MouthShapeCasing MouthShapeCasing { get; }

		public static MouthNaming Guess(IReadOnlyCollection<string> mouthNames) {
			string firstMouthName = mouthNames.First();
			if (mouthNames.Count == 1) {
				return firstMouthName == string.Empty
					? new MouthNaming(string.Empty, string.Empty, MouthShapeCasing.Lower)
					: new MouthNaming(
						firstMouthName.Substring(0, firstMouthName.Length - 1),
						string.Empty,
						GuessMouthShapeCasing(firstMouthName.Last()));
			}

			string commonPrefix = mouthNames.GetCommonPrefix();
			string commonSuffix = mouthNames.GetCommonSuffix();
			var mouthShapeCasing = firstMouthName.Length > commonPrefix.Length
				? GuessMouthShapeCasing(firstMouthName[commonPrefix.Length])
				: MouthShapeCasing.Lower;
			return new MouthNaming(commonPrefix, commonSuffix, mouthShapeCasing);
		}

		public string DisplayString {
			get {
				string casing = MouthShapeCasing == MouthShapeCasing.Upper
					? "<UPPER-CASE SHAPE NAME>"
					: "<lower-case shape name>";
				return $"\"{Prefix}{casing}{Suffix}\"";
			}
		}

		public string GetName(MouthShape mouthShape) {
			string name = MouthShapeCasing == MouthShapeCasing.Upper
				? mouthShape.ToString()
				: mouthShape.ToString().ToLowerInvariant();
			return $"{Prefix}{name}{Suffix}";
		}

		private static MouthShapeCasing GuessMouthShapeCasing(char mouthShape) {
			return char.IsUpper(mouthShape) ? MouthShapeCasing.Upper : MouthShapeCasing.Lower;
		}
	}

	public enum MouthShapeCasing {
		Upper,
		Lower
	}
}

File: extras/EsotericSoftwareSpine/src/main/kotlin/com/rhubarb_lip_sync/rhubarb_for_spine/MouthNaming.kt
package com.rhubarb_lip_sync.rhubarb_for_spine

import java.util.*

class MouthNaming(private val prefix: String, private val suffix: String, private val mouthShapeCasing: MouthShapeCasing) {

	companion object {
		fun guess(mouthNames: List<String>): MouthNaming {
			if (mouthNames.isEmpty()) {
				return MouthNaming("", "", guessMouthShapeCasing(""))
			}

			val commonPrefix = mouthNames.commonPrefix
			val commonSuffix = mouthNames.commonSuffix
			val firstMouthName = mouthNames.first()
			if (commonPrefix.length + commonSuffix.length >= firstMouthName.length) {
				return MouthNaming(commonPrefix, "", guessMouthShapeCasing(""))
			}

			val shapeName = firstMouthName.substring(
				commonPrefix.length,
				firstMouthName.length - commonSuffix.length)
			val mouthShapeCasing = guessMouthShapeCasing(shapeName)
			return MouthNaming(commonPrefix, commonSuffix, mouthShapeCasing)
		}

		private fun guessMouthShapeCasing(shapeName: String): MouthShapeCasing {
			return if (shapeName.isBlank() || shapeName[0].isLowerCase())
				MouthShapeCasing.Lower
			else
				MouthShapeCasing.Upper
		}
	}

	fun getName(mouthShape: MouthShape): String {
		val name = if (mouthShapeCasing == MouthShapeCasing.Upper)
			mouthShape.toString()
		else
			mouthShape.toString().toLowerCase(Locale.ROOT)
		return "$prefix$name$suffix"
	}

	val displayString: String get() {
		val casing = if (mouthShapeCasing == MouthShapeCasing.Upper)
			"<UPPER-CASE SHAPE NAME>"
		else
			"<lower-case shape name>"
		return "\"$prefix$casing$suffix\""
	}
}

enum class MouthShapeCasing {
	Upper,
	Lower
}